### PR TITLE
zktrie part3: include zktrie witness in block trace; add demo for generating witness data for mpt circuit

### PR DESCRIPTION
**Pull request by noel2004**
This code would allow some witness data used by mpt circuit being added into the block result. It would be controlled by cmd paramete `trace.mptwitness=<output scheme>`. The `output scheme` would be a number indicating the order of ops sequence (see below) and by default it is 0, which means there is no any output.

For the storage state transistion inside a block, there would be many possible op sequence for mpt circuit to provide. For example, it would be the the "natrual" sequence, which collect ops in the order they were involved by the op codes executed in each tx; (the `output scheme` is 1.) or the "state" sequence like they are ordered in the rw table. (the `output scheme` is 2.) 

Notice we consider the code here would be just a "demo" because to output an op sequence which would be completly consistent with the purposed scheme which connect storage and state circuit. The logic are currently implied in busmapping and it should not be duplicated in l2geth.

### CODE REVIEW

**Comment by lispc** [comment]
we should gen_zktrie_witness_from_block_result to goroller later but now adding zktrie witness into blockresult is the most convenient way for the overall procedure at the cost of some computing overhead on l2geth

**Comment by lispc** [comment]
@noel2004 it this branch still WIP?

**Comment by noel2004** [comment]
> @noel2004 it this branch still WIP?

Yes. Another orderer, which would generate ops with the same order in rw table, is underway. It would be put on today.

**Comment by noel2004** [comment]
Current the witness can be genreated with order respect to rw table. I may have made some redudant "read" op for each account being accessed for do not consider more knowledge on each operation inside evm. However, it would sitll work well if we have adopted [the new lookup scheme currently under discussed](https://github.com/privacy-scaling-explorations/zkevm-specs/issues/217).

To enable this feature, use argument `--trace.mptwitness=2` when launching geth

**Review by lispc — DISMISSED** [review]

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -255,6 +255,12 @@ var (
 		Usage: "Handle the latest several blockResults",
 		Value: ethconfig.Defaults.TraceCacheLimit,
 	}
+	// mpt witness settings
+	MPTWitness = cli.IntFlag{
+		Name:  "trace.mptwitness",
+		Usage: "Output witness for mpt circuit with Specified order (default = no output, 1 = by executing order",
+		Value: ethconfig.Defaults.MPTWitness,
+	}
```
Comment (0xmountaintop) [inline cmd/utils/flags.go]: I think we should move this after https://github.com/scroll-tech/go-ethereum/blob/zkrollup/internal/debug/flags.go#L93 (also move `TraceCacheLimit`)

File: cmd/geth/main.go [hunk 2]
```diff
@@ -201,6 +201,11 @@ var (
 		utils.MetricsInfluxDBBucketFlag,
 		utils.MetricsInfluxDBOrganizationFlag,
 	}
+
+	traceFlags = []cli.Flag{
+		utils.TraceCacheLimit,
+		utils.MPTWitness,
+	}
```
Comment (0xmountaintop) [inline cmd/geth/main.go]: I think we should combine it into https://github.com/scroll-tech/go-ethereum/blob/zkrollup/internal/debug/flags.go#L109

File: common/hexutil/hexutil.go [hunk 3]
```diff
@@ -64,6 +64,11 @@ func Decode(input string) ([]byte, error) {
 	if !has0xPrefix(input) {
 		return nil, ErrMissingPrefix
 	}
+	// better compatible with odd size string
+	if len(input)%2 != 0 {
+		input = "0x0" + input[2:]
+	}
+
```
Comment (0xmountaintop) [inline common/hexutil/hexutil.go:71]: If we can go here, `has0xPrefix` must be `true`, it makes no sense to add `0x0` here

File: core/blockchain.go [hunk 4]
```diff
@@ -134,6 +135,7 @@ type CacheConfig struct {
 	SnapshotLimit       int           // Memory allowance (MB) to use for caching snapshot entries in memory
 	Preimages           bool          // Whether to store preimage of trie key to the disk
 	TraceCacheLimit     int
+	MPTWitness          int // How to generate witness data for mpt circuit, 0: nothing, 1: natural
```
Comment (0xmountaintop) [inline core/blockchain.go:138]: make use of iota

File: common/hexutil/hexutil.go (hunk 3 above)
Comment (noel2004) [reply common/hexutil/hexutil.go:71]: Here I wish to handle the case that some hex printer never output a leading 0 so there would be hex string with odd bytes, like "0x12345".

The old code would complain in this cases so here the input would be corrected to "0x012345"

File: core/blockchain.go (hunk 4 above)
Comment (noel2004) [reply core/blockchain.go:138]: it means we should use a string type here?

Comment (0xmountaintop) [reply core/blockchain.go:138]: for example

```
type MPTWitnessType uint

const (
	MPTWitnessNothing MPTWitnessType = iota
	MPTWitnessNatural
)
```

Then `MPTWitnessNothing` will be `0` (in terms of value), `MPTWitnessNatural` will be `1`

Comment (0xmountaintop) [reply core/blockchain.go:138]: Then line 152

```
	MPTWitness:      0,
```

can be 
```
	MPTWitness:      MPTWitnessNothing,
```

File: trie/zkproof/writer.go [hunk 5]
```diff
@@ -0,0 +1,729 @@
...
+	}
+}
+
+func copyAccountState(st *types.AccountWrapper) *types.AccountWrapper {
+	return &types.AccountWrapper{
+		Nonce:    st.Nonce,
+		Balance:  st.Balance,
+		CodeHash: st.CodeHash,
+		Address:  st.Address,
+		Storage:  st.Storage,
```
Comment (0xmountaintop) [inline trie/zkproof/writer.go]: can it really copy the AccountState instead of re-using the pointers of `Balance` & `Storage`?

Comment (noel2004) [reply trie/zkproof/writer.go]: So you prefer to we should have a deep copy of these 2 fields？

Comment (0xmountaintop) [reply trie/zkproof/writer.go]: If using the pointer is fine then it's fine.

TBH I don't know what is desired.

File: trie/zkproof/writer.go [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+			contractStack[sl] = currentContract
+			currentContract = callEnterAddress
+		} else if sl > sLog.Depth {
+			logStack = logStack[:sl-1]
+			currentContract = contractStack[sLog.Depth]
+			resumePos := logStack[len(logStack)-1]
+			calledLog := logs[resumePos]
+
+			//no need to handle fail calling
+			if calledLog.ExtraData == nil || !calledLog.ExtraData.CallFailed {
```
Comment (0xmountaintop) [inline trie/zkproof/writer.go]: I think
```suggestion
			if !(calledLog.ExtraData != nil && calledLog.ExtraData.CallFailed) {
```

is more readable

File: trie/zkproof/writer.go (hunk 5 above)
Comment (noel2004) [reply trie/zkproof/writer.go]: AccountState objects is expected to not be changed, that is why we need this function when produce another state. But I admit that the `balance` is prone to be mutated accidently so deep copy it should be much safer. For `Storage`, it is highly impossible to be changed inside any codes so I think using pointer would be ok?

File: internal/debug/flags.go [hunk 7]
```diff
@@ -117,6 +131,23 @@ func init() {
...
+type TraceConfig struct {
+	TracePath string
+	// Trace option
+	TraceCacheLimit int
+	MPTWitness      int
+}
+
+func ConfigTrace(ctx *cli.Context) *TraceConfig {
+	cfg := new(TraceConfig)
+	cfg.TracePath = ctx.GlobalString(traceFlag.Name)
```
Comment (HAOYUatHZ) [inline internal/debug/flags.go:144]:
```suggestion
	if ctx.GlobalIsSet(traceFlag.Name) {
		cfg.TracePath = ctx.GlobalString(traceFlag.Name)
	}
```

**Comment by 0xmountaintop** [comment]
Overall looks good to me now.

Just some nitpicks.

**Review by 0xmountaintop — APPROVED** [review]

**Pull request merged**